lib/foreman/renderer.rb

```diff
@@ -1,390 +1,125 @@
+                                                  subjects: subjects,
+                                                  params: params)
+        renderer_class.new(template_scope).render(variables: variables)
       end
 
-      @provisioning_type = @host.is_a?(Hostgroup) ? 'hostgroup' : 'host'
-
-      # force static network configuration if static http parameter is defined, in the future this needs to go into the GUI
-      @static = !params[:static].empty?
-
-      # this is sent by the proxy when the templates feature is enabled
-      # and is needed to direct the host to the correct url. without it, we increase
-      # latency by requesting the correct url directly from the proxy.
-      @template_url = params['url']
-    end
-
-    def global_setting(name, blank_default = nil)
-      raise(FilteredGlobalSettingAccessed, _('Global setting "%s" is not accessible in safe-mode') % name) if Setting[:safemode_render] && !ALLOWED_GLOBAL_SETTINGS.include?(name.to_sym)
-      setting = Setting.find_by_name(name.to_sym)
-      (setting.settings_type != "boolean" && setting.value.blank?) ? blank_default : setting.value
-    end
-
-    private
-
-    def check_host
-      raise HostUnknown, _('This templates requires a host to render but none was specified') if @host.nil?
-    end
-
-    def check_host_param(name)
-      check_host
-      raise(HostParamUndefined, _('Parameter %{name} is not set for host %{host}') % { :name => name, :host => @host }) unless @host.params.key?(name)
-    end
-
-    # takes variable names array and loads instance variables with the same name like this
-    # { :name => @name, :another => @another }
-    def allowed_variables_mapping(variable_names)
-      variable_names.reduce({}) do |mapping, var|
-        mapping.update(var => instance_variable_get("@#{var}"))
+      def render_template_to_tempfile(template, prefix, options = {})
+        file = ""
+        Tempfile.open(prefix, Rails.root.join('tmp')) do |f|
+          f.print(render_template(template: template))
+          f.flush
+          f.chmod options[:mode] if options[:mode]
+          file = f
+        end
+        file
       end
-    end
-
-    def alterator_attributes
-      @mediapath   = @host.operatingsystem.mediumpath @host
-      @mediaserver = URI(@mediapath).host
-      @metadata    = params[:metadata].to_s
-    end
 
-    def jumpstart_attributes
-      if @host.operatingsystem.supports_image && @host.use_image
-        @install_type     = "flash_install"
-        # We have an individual override for the host's image file
-        @archive_location = @host.image_file ? @host.image_file : @host.default_image_file
-      else
-        @install_type = "initial_install"
-        @system_type  = "standalone"
-        @cluster      = "SUNWCreq"
-        @packages     = "SUNWgzip"
-        @locale       = "C"
+      def template_scope_class
+        @template_scope_class ||= Foreman::Renderer::DatabaseTemplateScope
```

Comment:
We discussed this internally and don't want to provide the full URL to a snipped.
The way we designed our plugin is, that it downloads a .tar.gz from a git repo and looks up the snipped by filename in that repo.
But maybe we can just override the method from core in the plugins template scope.

Reply:
So you don't store the templates in the database but rather in a tarbal? I'm wondering if scope is the correct word or storage/store is a better description.

Reply:
> So you don't store the templates in the database but rather in a tarbal?

Yep, that's the idea.

> I'm wondering if scope is the correct word or storage/store is a better description.

Actually, we'll have both: A template scope (that overrides snippet macros etc.) and also a storage implementation.

Reply:
Interesting use case, I only wonder if template rendering is a good time for this, note that this would happen also when you just preview the template. But if tarball is small, perhaps it's fine. I'm just curious why you don't want to wrap existing `snippet` macro in another snippet, something like

tarball_snippet.rb:
```
wget https://example.com/<%= @name -%>.tar
tar xf <%= @name -%>.tar
snippet('<%= @name %>')
```

then you'd just call `snippet 'tarball_snippet', :variables => { :name => 'real_snippet_name' }` (snippets can be parametrized for few versions)

I mean this is the way, how users can define their own functions without modifying code in core

or if you are already having a plugin that would use it, you can simply add your own macros and use that instead of built-in snippet

and if this is because of templates storage, we probably don't need variable scope classes? or do you plan to store them externally as well?

Reply:
The concept of splitting it from the database would make it easier to move out of Foreman into another service (smart-proxy, standalone). Just provide the context.